collections/map.go

```diff
@@ -0,0 +1,134 @@
+
+import (
+	"context"
+	"fmt"
+
+	storetypes "github.com/cosmos/cosmos-sdk/store/types"
+)
+
+// NewMap returns a Map given a StoreKey, a Prefix and the relative value and key encoders.
+func NewMap[K, V any](
+	sk storetypes.StoreKey, prefix Prefix,
+	keyEncoder KeyCodec[K], valueEncoder ValueCodec[V],
+) Map[K, V] {
+	return Map[K, V]{
+		kc:     keyEncoder,
+		vc:     valueEncoder,
+		sk:     sk,
+		prefix: prefix.Bytes(),
+	}
+}
+
+// Map represents the basic collections object.
+// It is used to map arbitrary keys to arbitrary
+// objects.
+type Map[K, V any] struct {
+	kc KeyCodec[K]
+	vc ValueCodec[V]
+
+	sk     storetypes.StoreKey
+	prefix []byte
+}
+
+// Set maps the provided value to the provided key in the store.
+// Errors with ErrEncoding if key or value encoding fails.
+func (m Map[K, V]) Set(ctx context.Context, key K, value V) error {
+	keyBytes, err := m.encodeKey(key)
+	if err != nil {
+		return err
+	}
+
+	valueBytes, err := m.vc.Encode(value)
+	if err != nil {
+		return fmt.Errorf("%w: value encode: %s", ErrEncoding, err) // TODO: use multi err wrapping in go1.20: https://github.com/golang/go/issues/53435
+	}
+
+	store, err := m.getStore(ctx)
+	if err != nil {
+		return err
+	}
+	store.Set(keyBytes, valueBytes)
+	return nil
+}
+
+// Get returns the value associated with the provided key,
+// errors with ErrNotFound if the key does not exist, or
+// with ErrEncoding if the key or value decoding fails.
+func (m Map[K, V]) Get(ctx context.Context, key K) (V, error) {
+	keyBytes, err := m.encodeKey(key)
+	if err != nil {
+		var v V
+		return v, err
+	}
+
+	store, err := m.getStore(ctx)
+	if err != nil {
+		var v V
+		return v, err
+	}
+	valueBytes := store.Get(keyBytes)
+	if valueBytes == nil {
+		var v V
+		return v, fmt.Errorf("%w: key '%s' of type %s", ErrNotFound, m.kc.Stringify(key), m.vc.ValueType())
+	}
+
+	v, err := m.vc.Decode(valueBytes)
+	if err != nil {
+		var v V
+		return v, fmt.Errorf("%w: value decode: %s", ErrEncoding, err) // TODO: use multi err wrapping in go1.20: https://github.com/golang/go/issues/53435
+	}
+	return v, nil
+}
+
+// Has reports whether the key is present in storage or not.
+// Errors with ErrEncoding if key encoding fails.
+func (m Map[K, V]) Has(ctx context.Context, key K) (bool, error) {
+	bytesKey, err := m.encodeKey(key)
+	if err != nil {
+		return false, err
+	}
+	store, err := m.getStore(ctx)
+	if err != nil {
+		return false, err
+	}
+	return store.Has(bytesKey), nil
+}
+
+// Remove removes the key from the storage.
+// Errors with ErrEncoding if key encoding fails.
+// If the key does not exist then this is a no-op.
+func (m Map[K, V]) Remove(ctx context.Context, key K) error {
+	bytesKey, err := m.encodeKey(key)
+	if err != nil {
+		return err
+	}
+	store, err := m.getStore(ctx)
+	if err != nil {
+		return err
+	}
+	store.Delete(bytesKey)
+	return nil
+}
+
+func (m Map[K, V]) getStore(ctx context.Context) (storetypes.KVStore, error) {
+	provider, ok := ctx.(StorageProvider)
```

Comment:
A context.Context should never _be_ a StorageProvider in this sense. It might _have_ a StorageProvider, accessible by e.g. context.Value(...).

Reply:
Unfortunately, `sdk.Context` implements `context.Context`, and also has a `KVStore` method, but you can defer this comment into a different PR, since for performance reasons, this will not be a viable path to not depend on `cosmos/cosmos-sdk/`, ref: https://github.com/cosmos/cosmos-sdk/pull/14203/files#

Reply:
https://github.com/cosmos/cosmos-sdk/pull/14203

>  NOTE: Uses pointer receiver to save on execution time

Cool! Where can I see the benchmark data that demonstrates the improvement from this change?